### DISCUSSION TITLE
Default item selection in Carousel Mode?

### DISCUSSION THREAD

**Discussion by bernarma** [missionpinball/mpf#1825] [Q&A] [answered]
I would like to extend the behavior of the built-in Carousel mode to support having a default value provided as a mode setting/parameter (which could be either a dynamic value or a static string value that exists in the list).

I have extended the implementation so that it is fully working but I would like to know:

1. Is there a way that this can be achieved with the current implementation (requiring no modification)?
2. is there a reason why this wasn't included (i.e. bad practice or should be done another way or it wasn't needed at the time)?

Many thanks for the response.

Note: I can submit a pull request if you would like to get this additional feature incorporated into the main code base.

**Comment by avanwinkle** [comment]
On the one hand, it depends whether you're asking about MPF-MC or GMC. On the other hand, no, there's no way it can be achieved with the current implementation :)

The carousel is managed by MPF as an array of items and rendered in the media controller through events. With MPF being the source of truth and the "selection" being an index of the array of items, there is always _something_ selected.

If the desire is to conditionally choose which item is selected first, conditional events in the carousel `selectable_items` list will control that. Or you could trigger an event to select an item when/as the mode starts to adjust the `_highlighted_item_index` to the carousel item you prefer.

What situation are you facing where a default value for the carousel is the solution?

**Reply by bernarma** [reply]
To answer your first question, I am using GMC as my media controller. Since I am starting a new ruleset, I wanted to be using the latest and greatest ;-).

Now on to the use case... What I am trying to achieve is related to the selection of a mode at various points in the game (whether that be at the start of a players first or subsequent ball or once that mode has reached a particular state - i.e. it finishes or all objectives have been completed).

My goal is when asking the user to choose a new item from the list I want it to start from a item (i.e. the one that is related to the running mode). As a side-note, this item has the potential to change outside of the carousel component I am talking about and I am currently storing this value in the current_player.item_name property.

Hopefully this all makes sense. I am happy to look at your other options you proposed though to see if I can make it work without any modifications.

As for setting the _highlighted_item_index value, is this something I would do in a python code file alongside the mode or is there a yaml based way I am able to leverage?

**Comment by avanwinkle** [comment, marked as answer]
I did something exactly that in my first game, where the player could choose from a variety of missions. If they played but did not complete a mission, the first item in the carousel would always be the last mission they attempted.

https://github.com/avanwinkle/masseffect2/blob/master/modes/missionselect/code/missionselect.py#L90-L95

(That block is the only relevant section, all the rest of that code is for much more complex stuff)

**Reply by bernarma** [reply]
Thanks for showing me the example. It looks as though you are "ordering" the list so that the one you want selected will be the first item in the list. If that's the case, I may follow that approach as I am more inclined to stick to semi-out-of-the-box functionality than maintain a separate feature entirely for no added benefit.

Looking at the entirety of the file you shared, I did start originally extending Carousel but I was trying to alter the index as opposed to just changing the order of the list... I am going to try this and see how I go. Thanks once again.

**Reply by bernarma** [reply]
OK... So much more elegant!!!

```
from mpf.modes.carousel.code.carousel import Carousel

class SongSelect(Carousel):

    def mode_init(self):
        """Init: create a logger."""
        super().mode_init()

    def mode_start(self, **kwargs):
        """Mode start: build a list of available songs."""
        self._all_items = self._build_items_list()

        super().mode_start(**kwargs)

    def _build_items_list(self):
        player = self.machine.game.player
        index = self._all_items.index(player["selected_song"])
        return self._all_items[index:] + self._all_items[:index]
```

**Reply by avanwinkle** [reply]
Ah hah, so you essentially "rotate" the list so that the selected item is first but the selection list will always be in the same order. Fantastic! :D
